src/com/amazon/corretto/crypto/provider/AesGcmSpi.java

```diff
@@ -971,17 +971,26 @@ protected int engineDoFinal(final ByteBuffer input, final ByteBuffer output)
     return output.position() - initialPosition;
   }
 
-  private void checkOutputBuffer(final int inputLength, final byte[] output, final int outputOffset)
+  private void checkOutputBuffer(
+      final int inputLength, final byte[] output, final int outputOffset, final boolean doFinal)
       throws ShortBufferException {
-    if (inputLength < 0 || outputOffset < 0 || outputOffset > output.length) {
+    final int freeBufferSpace = output.length - outputOffset;
+    final int requiredBufferSpace =
+        doFinal ? engineGetOutputSize(inputLength) : getUpdateOutputSize(inputLength);
+    if (inputLength < 0 || outputOffset < 0) {
       throw new ArrayIndexOutOfBoundsException();
     }
-    if (output.length - outputOffset < getUpdateOutputSize(inputLength)) {
+    // Allow outputOffset to index into a 0-length empty array. We must trust that the rest of the
+    // code doesn't actually do this.
+    if (outputOffset > output.length || (outputOffset == output.length && output.length != 0)) {
```

Comment:
nit: might be good to have branch coverage on the output.length == 0 case.

Reply:
FWIW, this is exercised as the happy case for a lot of our doUpdate/decrypt tests and incidentally a part of the doFinal/decrypt paths.